File: backend/PulseMap/Domain/Location.cs
using PulseMap.Domain.Enums;

namespace PulseMap.Domain;

public class Location
{
    public required int Id { get; set; }
    public required double Latitude { get; set; }
    public required double Longitude { get; set; }
    public required string Name { get; set; } = string.Empty;
    public required string? Description { get; set; } = string.Empty;

    public required int? CreatorId { get; set; } = null;
    public User? Creator { get; set; } = null;

    public required Category Category { get; set; } = Category.NotSet;
    public List<Message>? Comments { get; set; } = [];

    public required DateTime ExpiresAt { get; set; }
    public required bool IsExpired { get; set; } = false;

    public List<User> Likes { get; set; } = [];
    public int? OwnerId { get; set; } = null;
    public User? Owner { get; set; } = null;

    public LikeStatus? LikeStatus { get; set; } = null;

    // Event clustering
    public int? EventId { get; set; } = null;
    public Event? Event { get; set; } = null;
    public float? EventAssignmentConfidence { get; set; } = null;
}

File: shared/src/types/location.ts
export interface Location {
  id: number;
  latitude: number;
  longitude: number;
  name: string;
  creator: SimplifiedUser;
  category: string;
  description?: string;
  messages: Message[];
  expiresAt: Date;
  isExpired: boolean;
  likesCount: number;
  isLikedByCurrentUser: boolean;
  requiresReview: boolean;
  owner?: SimplifiedUser;
  event?: SimplifiedEvent;
  eventAssignmentConfidence?: number;
  imageUrls?: string[];
}

export enum LocationCategory {
  NotSet = 'Not Set',
  Music = 'Music',
  Sport = 'Sport',
  Food = 'Food',
  Entertainment = 'Entertainment',
  Education = 'Education',
  Health = 'Health',
  Technology = 'Technology',
  Travel = 'Travel',
  Art = 'Art',
  Business = 'Business',
}

export interface Message {
  id: number;
  content: string;
  sentAt: Date;
  sender: SimplifiedUser;
  locationId: number;
  responses?: ResponseMessage[];
}

export interface ResponseMessage {
  id: number;
  sender: SimplifiedUser;
  content: string;
  sentAt: Date;
  locationId: number;
  parentMessageId: number;
}

export interface LocationLikesSummaryDTO {
  id: number;
  likesCount: number;
  toggledByUserId?: number | null;
  isNowLiked: boolean;
}

export interface LocationPostDTO {
  latitude: number;
  longitude: number;
  name: string;
  description?: string;
  creatorId: number;
  category: string;
  duration: string;
  ownerId?: number;
  imageUrls?: string[];
}

export interface LocationPutDTO {
  name: string;
  description?: string;
  category: string;
  imageUrls?: string[];
}

export interface MessagePostDTO {
  content: string;
  senderId: number;
  locationId: number;
}

export interface ResponseMessagePostDTO {
  content: string;
  senderId: number;
  messageId: number;
}

export interface CategoryDTO {
  id: number;
  name: string;
  slug: string;
  isActive: boolean;
  sortOrder: number;
}

export interface LocationRecommendationDTO {
  id: number;
  name: string;
  description: string;
  category: string;
  latitude: number;
  longitude: number;
  likesCount: number;
  score: number;
  reason: string;
}

export interface SimplifiedUser {
  id: number;
  firstName: string;
  lastName: string;
  username: string;
}

export interface SimplifiedEvent {
  id: number;
  name: string;
  requiresReview: boolean;
  confidenceScore: number;
}

export interface MapBounds {
  minLat: number;
  maxLat: number;
  minLng: number;
  maxLng: number;
}
